### PR TITLE
Ft nft mint quest

### PR DESCRIPTION
**Pull request by mubarak23**
- [ ] issue #104 
- [*] code change includes tests

### CODE REVIEW

**Comment by mubarak23** [comment]
am getting the following error when i ran the test
 `Got an exception while executing a hint: Hint Error: 0x4661696c656420746f20646573657269616c697a6520706172616d202331 ('Failed to deserialize param #1')
 `
@b-j-roberts what could be the cause of this base on the test i write

**Comment by ptisserand** [comment]
> am getting the following error when i ran the test `Got an exception while executing a hint: Hint Error: 0x4661696c656420746f20646573657269616c697a6520706172616d202331 ('Failed to deserialize param #1') ` @b-j-roberts what could be the cause of this base on the test i write

Hi, you declare the wrong contract [here](https://github.com/keep-starknet-strange/art-peace/pull/121/files#diff-bbbf8950c16088ad8ea126bf7048821b41b3e07f6ca455c1bd47baae373ec823R19) 
You have to call `declare("NFTMintQuest");` instead of `declare("CanvasNFT");`

**Comment by ptisserand** [comment]
> > Hi, you declare the wrong contract [here](https://github.com/keep-starknet-strange/art-peace/pull/121/files#diff-bbbf8950c16088ad8ea126bf7048821b41b3e07f6ca455c1bd47baae373ec823R19) You have to call `declare("NFTMintQuest");` instead of `declare("CanvasNFT");`
> 
> you are right here,
> 
> after changing it , am getting the following error
> 
> `Got an exception while executing a hint: Requested contract address ContractAddress(PatriciaKey(StarkFelt("0x000000000000000000000000000000000000000000000043616e7661734e4654"))) is not deployed.`

[`utils::NFT_CONTRACT()](https://github.com/keep-starknet-strange/art-peace/pull/121/files#diff-bbbf8950c16088ad8ea126bf7048821b41b3e07f6ca455c1bd47baae373ec823R22) doesn't deploy a NFT contract, it only returns an hardcoded contract address.

In your `is_claimable` implementation, you call [`get_nft_minter`](https://github.com/keep-starknet-strange/art-peace/pull/121/files#diff-bd55666dfd92537a1a166c3f6a49fd7611db8882d64efbdd8927eceb75c42de1R51).
So you need to declare&deploy a `CanvasNFT` before deploying your `NFTMintQuest` contract.

**Comment by mubarak23** [comment]
> > > Hi, you declare the wrong contract [here](https://github.com/keep-starknet-strange/art-peace/pull/121/files#diff-bbbf8950c16088ad8ea126bf7048821b41b3e07f6ca455c1bd47baae373ec823R19) You have to call `declare("NFTMintQuest");` instead of `declare("CanvasNFT");`
> > 
> > 
> > you are right here,
> > after changing it , am getting the following error
> > `Got an exception while executing a hint: Requested contract address ContractAddress(PatriciaKey(StarkFelt("0x000000000000000000000000000000000000000000000043616e7661734e4654"))) is not deployed.`
> 
> [`utils::NFT_CONTRACT()](https://github.com/keep-starknet-strange/art-peace/pull/121/files#diff-bbbf8950c16088ad8ea126bf7048821b41b3e07f6ca455c1bd47baae373ec823R22) doesn't deploy a NFT contract, it only returns an hardcoded contract address.
> 
> In your `is_claimable` implementation, you call [`get_nft_minter`](https://github.com/keep-starknet-strange/art-peace/pull/121/files#diff-bd55666dfd92537a1a166c3f6a49fd7611db8882d64efbdd8927eceb75c42de1R51). So you need to declare&deploy a `CanvasNFT` before deploying your `NFTMintQuest` contract.

so after i deploying the NFT_CANVAS contract, am now getting the following   


```
Failure data:
    Got an exception while executing a hint: Hint Error: Error in the called contract (0x0735d0f59362f724897cee0ef1d6ad97ece788dcc126fefb45656ff43d872698):
Error at pc=0:1679:
Got an exception while executing a hint: Requested contract address ContractAddress(PatriciaKey(StarkFelt("0x00000000000000000000000000000000000000004e46544d696e745175657374"))) is not deployed.
Cairo traceback (most recent call last):
Unknown location (pc=0:271)
Unknown location (pc=0:1241)
```
only `art_peace::tests::nft_quest::test_is_claimable` fails

File: onchain/src/tests/nft_quest.cairo [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    let is_claim = false;
+
+    assert(is_claim == test_is_claim, 'Cannot Claim Mint NFT Quest');
+}
+
+#[test]
+fn test_claim() {
+    let contract_address = deploy_nft_quest();
+    let dispatcher = IQuestDispatcher { contract_address };
```
Comment (b-j-roberts) [inline onchain/src/tests/nft_quest.cairo:67]: The quest dispatcher should be the art_peace contract. Please use the tests here as reference to fix the issues in this file. Most things should be the same : https://github.com/keep-starknet-strange/art-peace/blob/main/onchain/src/tests/template_quest.cairo

**Review by b-j-roberts — CHANGES_REQUESTED** [review]

File: onchain/src/quests/nft_quest.cairo [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+            if token_miner != user {
+                return false;
+            }
+
+            return true;
+        }
+
+        fn claim(ref self: ContractState, user: ContractAddress, calldata: Span<felt252>) -> u32 {
+            if get_caller_address() != self.CanvasNFT.read() {
```
[the remark this answers is not in the document]
Comment (mubarak23) [reply onchain/src/quests/nft_quest.cairo:61]: i see this

File: onchain/src/tests/nft_quest.cairo [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+use snforge_std as snf;
+use snforge_std::{declare, CheatTarget, ContractClassTrait};
+
+const reward_amt: u32 = 18;
+
+fn SINGLE_CALLDATA() -> Span<felt252> {
+    array![1].span()
+}
+
+fn deploy_nft_contract() -> ContractAddress {
```
[the remark this answers is not in the document]
Comment (mubarak23) [reply onchain/src/tests/nft_quest.cairo:18]: if am implementing almost exactly what was done inside `test/template_quest.cairo` then we don't need this function

File: onchain/src/tests/nft_quest.cairo (hunk 1 above)
Comment (mubarak23) [reply onchain/src/tests/nft_quest.cairo:67]: Most thing should be the same, but we are switching `deploy_template_quest` with   `deploy_nft_quest` ?

> The quest dispatcher should be the art_peace contract. Please use the tests here as reference to fix the issues in this file. Most things should be the same : https://github.com/keep-starknet-strange/art-peace/blob/main/onchain/src/tests/template_quest.cairo

**Pull request merged**